# Python standard library
import json
from glob import glob
from os.path import join
from os import listdir
from pathlib import Path
from snakemake.workflow import workflow as wf_api
from snakemake.utils import R
from scripts.common import (
    allocated,
    provided, 
    references,
    str_bool
)


configfile: "config.yaml"

# Get list of BAM files from the specified directory
bam_dir = config.get("bam_dir", "bams")
SAMPLES = [Path(f).stem for f in glob(f"{bam_dir}/*.bam")]

# Validate SAMPLES
if not SAMPLES:
    raise ValueError(f"No BAM files found in {bam_dir}")

# Select genome-specific parameters
GENOME = config["genome"]
if GENOME not in config:
    raise ValueError(f"Genome {GENOME} not defined in config.yaml")
GENOME_PARAMS = config[GENOME]

# Define outputs
outputs = []

# Coverage
if GENOME_PARAMS.get("intervals"):
    outputs.extend(expand("coverage/{sample}_coverage.bed", sample=SAMPLES))
else:
    print("Warning: Skipping coverage rule due to missing intervals")

# Fragment length bins
outputs.extend(expand("frag_length_bins/{sample}_frag_bin{bin_size}.tsv", sample=SAMPLES, bin_size=[config['bin_size']]))

# Fragment length intervals
if GENOME_PARAMS.get("intervals"):
    outputs.extend(expand("frag_length_intervals/{sample}_frag_interval.bed", sample=SAMPLES))
else:
    print("Warning: Skipping frag_length_intervals rule due to missing intervals")

# End motifs
if GENOME_PARAMS.get("ref2bit"):
    outputs.extend(expand("end_motifs/{sample}_endmotif.tsv", sample=SAMPLES))
else:
    print("Warning: Skipping end_motifs rule due to missing ref2bit")

# Interval end motifs
if GENOME_PARAMS.get("ref2bit") and GENOME_PARAMS.get("intervals"):
    outputs.extend(expand("interval_end_motifs/{sample}_endmotif_interval.tsv", sample=SAMPLES))
else:
    print("Warning: Skipping interval_end_motifs rule due to missing ref2bit or intervals")

# MDS
if GENOME_PARAMS.get("ref2bit"):
    outputs.extend(expand("mds/{sample}_mds.tsv", sample=SAMPLES))
else:
    print("Warning: Skipping mds rule due to missing ref2bit")

# DELFI
if GENOME_PARAMS.get("chrom_sizes") and GENOME_PARAMS.get("ref2bit") and GENOME_PARAMS.get("intervals"):
    outputs.extend(expand("delfi/{sample}_delfi.bed", sample=SAMPLES))
else:
    print("Warning: Skipping delfi rule due to missing chrom_sizes, ref2bit, or intervals")

# WPS
if GENOME_PARAMS.get("tss"):
    outputs.extend(expand("wps/{sample}_wps_out_tss.bw", sample=SAMPLES))
else:
    print("Warning: Skipping wps rule due to missing tss")

# Adjusted WPS
if GENOME_PARAMS.get("tss") and GENOME_PARAMS.get("chrom_sizes") and GENOME_PARAMS.get("tss_interval"):
    outputs.extend(expand("adjust_wps/{sample}_wps_out_tss_adjusted.bw", sample=SAMPLES))
else:
    print("Warning: Skipping adjust_wps rule due to missing tss, chrom_sizes, or tss_interval")

# Cleavage profile
if GENOME_PARAMS.get("tss") and GENOME_PARAMS.get("chrom_sizes") and GENOME_PARAMS.get("tss_interval"):
    outputs.extend(expand("cleavage_profile/{sample}_cleavage_profile_tss.bw", sample=SAMPLES))
else:
    print("Warning: Skipping cleavage_profile rule due to missing tss, chrom_sizes, or tss_interval")

# Aggregation rules
if GENOME_PARAMS.get("tss_interval"):
    if GENOME_PARAMS.get("tss"):
        outputs.extend(expand("wps/{sample}_wps_out_tss_aggr.wig", sample=SAMPLES))
    else:
        print("Warning: Skipping agg_wps rule due to missing tss")
    if GENOME_PARAMS.get("tss") and GENOME_PARAMS.get("chrom_sizes") and GENOME_PARAMS.get("tss_interval"):
        outputs.extend(expand("adjust_wps/{sample}_wps_out_tss_adj_aggr.wig", sample=SAMPLES))
    else:
        print("Warning: Skipping agg_adjust_wps rule due to missing tss, chrom_sizes, or tss_interval")
    if GENOME_PARAMS.get("tss") and GENOME_PARAMS.get("chrom_sizes") and GENOME_PARAMS.get("tss_interval"):
        outputs.extend(expand("cleavage_profile/{sample}_cleavage_profile_aggr.wig", sample=SAMPLES))
    else:
        print("Warning: Skipping agg_cleavage_profile rule due to missing tss, chrom_sizes, or tss_interval")


rule all:
    input:
        outputs


# Import rules 
include: join("rules", "common.smk")
include: join("rules", "hooks.smk")
